### DISCUSSION TITLE
Ongoing developments

### DISCUSSION THREAD

**Discussion by pthom** [pthom/imgui_bundle#49] [Development Corner]
Hi @dcnieho :

I just opened this discussion so that we can exchange about ongoing development in a dedicated zone. I hope it will be OK for you.

> I'll send the make ImVec2/4 iterable patch later tonight.

Superb!

**Comment by pthom** [comment]
I need to inform you about important and fruitful changes I just applied: see 1d985954b5f444113

By changing the generator option `options.fn_exclude_non_api = False` (and also correcting the generator as a consequence 🫤), I added lots of new features, and they seem quite useful).
Look at [the diff](https://github.com/pthom/imgui_bundle/commit/1d985954b5f444113016ff64f094ec64b6feb1db) especially on `bindings/imgui_bundle/implot.pyi` and `bindings/imgui_bundle/implot.pyi`

You will see that we got `ImVec2.__getitem__` for free, + `ImLerp`, `ImClamp`, `Contains` and other goodies. 

This code now works:

````python
    vec2 = ImVec2(1, 2)
    x = vec2[0]
    y = vec2[1]
 ````

ImVec2 is still not iterable, no `__iter__` can be generated from C++.

**Reply by dcnieho** [reply]
Hmm, so i tried making `ImVec2` iterable, but failed. One way to make a type iterable is to define its `__len__` and `__getitem__` functions.
like so:
```
class myvec:
    def __init__(self):
        self.x=1
        self.y=2
    def __len__(self):
        return 2
    def __getitem__(self,index):
        if index==0:
            return self.x
        elif index==1:
            return self.y
        raise IndexError
# then all this works
a=myvec()
b=[*a]
# and you can also iterate over it
for p in a:
    print(p)
```

So i tried to define `__len__` in the pybind wrapper:
```
pyClassImVec2.def("__len__", [](const ImVec2& self) -> size_t {
        return 2;
    });
```

that works ok, `len(imgui.ImVec2(1,2))` returns 2. But when i then try a star expression or to iterate with for in, this happens:
```
a=imgui.ImVec2(1,2)
for i in a:
    print(i)
# after printing 1 and 2, i then get:
RuntimeError: IM_ASSERT( idx <= 1 )   ---   imgui.h:280
```
I get the same assertion with the star expression. This is the same you would get if you request e.g. a[2], which is non-existent of course.

**Reply by dcnieho** [reply]
Wait, more testing reveals what the problem is! If `__getitem__` runs out of data, IndexError should be raised so Python knows whats up. ImGui asserts and thus throws something else, no good.
`__len__` is not needed to make the class iterable. It is needed to make `reversed(class_object)` possible.

So, you may have to special-case generation for `__getitem__` functions if thats possible. I modified by hand to see if it works:
```
.def("__getitem__",
	    [](ImVec2 self, size_t idx) {
	    try {
	        return self[idx];
	    }
	    catch (...) {
                throw std::out_of_range("");
            }
    },
            py::arg("idx"),
            "(private API)\n\n We very rarely use this [] operator, the assert overhead is fine.")
```
With `__len__` additionally defined, `for i in reversed(imgui.ImVec2(1,2)):` works too.
I guess that is not something you can autogenerate. But it does work to define it like this. How about we disable specifically generated `__getitem__`s, see which disappear and add code like this by hand? Or can we manually opt-in a few classes somehow to have these additional methods (also `__len__` if there is a good way of figuring out the number of elements) generated for them?

**Reply by dcnieho** [reply]
Or not hacky in that it only catches indexing errors:
```
.def("__getitem__",
    [](ImVec2 self, size_t idx) {
        if (i >= 2) {
            throw py::index_error();
        }
	return self[idx];
    },
    py::arg("idx"),
    "(private API)\n\n We very rarely use this [] operator, the assert overhead is fine."
)
```

**Reply by dcnieho** [reply]
Ok, so those are dead ends. But this code, adapted from pybind testing code, does the trick:
```
    pyClassImVec2.def("__iter__", [](const ImVec2& self) {
            return py::make_iterator(&self.x, &self.x+2);
        },
        py::keep_alive<0, 1>() /* Essential: keep object alive while iterator exists */
    );
```
`__iter__` is preferred over `__getitem__`  if available, and thus sidesteps the problems we're having with that one.

**Reply by pthom** [reply]
What about a good old pure python [monkey patch](https://en.wikipedia.org/wiki/Monkey_patch) in __init__.py (and maybe also __init__.pyi)?

````python
def _iter_vec2(vec: ImVec2):
    yield vec.x
    yield vec.y

ImVec2.__iter__ = _iter_vec2

````

Beware I wrote this without testing it

**Reply by dcnieho** [reply]
I can do either, you may vote :)

**Reply by dcnieho** [reply]
(assuming both work)

**Reply by dcnieho** [reply]
I'll just do it on the C++ side, that works well

**Reply by pthom** [reply]
> I can do either, you may vote :)

I vote for three lines of python if it works. It is much more maintainable.

**Reply by dcnieho** [reply]
Ah hmm, i just tunnel-focused on delivering what i set out to and went to bed, didn't see this until now. Hope the C++ PR is good for you too.

**Comment by dcnieho** [comment]
Cool! Will check the diff in detail, and see if i can make it iterable, may be just another missing function.

**Reply by pthom** [reply]
Also see 2a448167 which adds warnings for the final user.

**Reply by dcnieho** [reply]
very nice stuff!

**Comment by pthom** [comment]
You will be happy: 4f4d25ed10c061ab636db8a7d8ebab612c764302 : ImGuiStyle.Colors is now bound as an python list/array 

Beware, I forked imgui_toggle (and had to update it). You may need to `git submodule sync`

**Reply by dcnieho** [reply]
Super! will test right away

**Reply by dcnieho** [reply]
It doesn't work for me. If i set a color through the array, the value is not updated. The interface doesn't change color, and if i get the color again, it is not my new value.

But see my comment here, there is (perhaps) a more general way to fix this than a patch? https://github.com/pthom/imgui_bundle/discussions/38#discussioncomment-4373476

Not high priority. 

It does make me wonder: does your generator output what things it doesn't wrap? Would be good to look through to see if some things should be wrapped, and if there are patterns/low hanging fruit where support for one case nets a lot (like the linked discussion)

**Reply by pthom** [reply]
> It does make me wonder: does your generator output what things it doesn't wrap?

Yes it does

**Comment by pthom** [comment]
> It doesn't work for me. If i set a color through the array, the value is not updated. The interface doesn't change color, and if i get the color again, it is not my new value.

This is strange. See http://www.giphy.com/gifs/IzoZ6z0xCGUgm16i8R

This is with the sandbox app: demos_python/sandbox_app.py

<img width="406" alt="image" src="https://user-images.githubusercontent.com/7694091/206931555-c5f50b11-2553-42ca-bbb5-830f53d49a61.png">

<img width="612" alt="image" src="https://user-images.githubusercontent.com/7694091/206931592-f21f2ada-c2cf-4593-a5e8-53f465b7a6ef.png">

**Reply by pthom** [reply]
Oh, I get it this is when you change the whole ImVec4 at a time.

Well, this will be for another day. I'm gonna call it a day, and you should to :-) Thanks a lot.

**Comment by pthom** [comment]
Hi Dee,

I merged your to PRs this morning. Yesterday I took a day as I was very tired.

Also, see 567eb106: I removed the access to ImGuiStyle.colors as an array, as it was broken.
Temporarily, we have`ImGuiStyle.color_(i)` and ImGuiStyle.set_color_(i)`(note the trailing underscore, which leaves an open possibility for a port as an array of something close)

With your cast ImVec2 from list, imgui.color_edit_4 now works perfectly, see sandbox_app.py

**Reply by dcnieho** [reply]
Thanks for that, glad they were useful! Of course man, you have been making some late nights i saw. Do not feel pressure from me.

I saw that commit indeed, will give it a test soonish.

**Comment by pthom** [comment]
@dcnieho 

Please see #52 where I opened a discussion  concerning array-like access to C array members. There is an initial study to do about the feasibility if you feel like investigating :-)

Concerning the generator and your attempts to compile it, it was never compiled under windows. Before going further, I'd like to discuss this with you by telephone (and may be help a little by remote access). I could be available at 3PM or 5PM.

**Reply by dcnieho** [reply]
Hmm, nice start! Not quite there yet in that ideally the machinery would be invisible to the end user, who just see an array. I'd love your help getting the generator to work, but i can't today, will contact you over email.

**Comment by pthom** [comment]
@dcnieho : warning concerning the latest commits to the dev and main branch.

I reorganized all the external libraries folder structure: see https://github.com/pthom/imgui_bundle/commit/867343400843889baa0574e268d49342932cb6d6

The external folder structure now looks like this:
<img width="320" alt="image" src="https://user-images.githubusercontent.com/7694091/210281983-a53efbb6-e586-4dca-a211-d9165a5cfe5b.png">

- Each library original source is stored inside external/{lib_name}/{lib_name}
- The bindings scripts + cpp pydef are stored inside  external/{lib_name}/bindings
- The generation scripts are stored inside external/bindings_generation

(The intent was to try to group the sources by library as much as possible).

Also see more comprehesive doc about bindings inside https://github.com/pthom/imgui_bundle/blob/main/external/_doc/Howto_Bindings.md

As a consequence, git will *very probably* choke when you run git update.

Your solutions are 
* either to re-clone the repo from scratch
* or suppress external/ and .git/modules/external/ before updating the repo

(this will happen only once)

thanks!

**Comment by dcnieho** [comment]
Indeed, it choked. So i recloned, seemed easiest. Current dev HEAD built
fine with pip install .
